### PR TITLE
Dictionary lookup interface

### PR DESCRIPTION
**Pull request by share-with-me**
The interface:

screenshot from 2017-08-13 16-14-19

### CODE REVIEW

**Comment by share-with-me** [comment]
Um, @sushain97 , it gives a `no-unused-vars` error for `currentLookupRequest` but I do use it in the `dictionary lookup` request.

**Comment by sushain97** [comment]
> Um, @sushain97 , it gives a no-unused-vars error for currentLookupRequest but I do use it in the dictionary lookup request.

You **write** to it but never **read** it. **use** == **read**.

**Comment by share-with-me** [comment]
> You write to it but never read it. use == read.

Now I get it :)

In the current log, it says, `apyRequestTimeout` not defined but I have included it in the `global` as well as the `import` in the file. This is the only error. It could be reviewed after this :D

**Comment by sushain97** [comment]
> but I have included it in the global as well as the import in the file.

Not in the right file?

**Comment by share-with-me** [comment]
> Not in the right file?

I had added to the file in `build`. Fixed it now :D

File: assets/css/style.css [hunk 1]
```diff
@@ -283,6 +283,28 @@ html[dir='rtl'] #srcLanguages {
...
+
+#dictTranslation {
+    color: #737373;
+}
+
+#currentWordForLookup {
+    color: #000;
+}
+
+#dynamicLocalization {
```
Comment (sushain97) [inline assets/css/style.css]: This should be more specific about what it actually refers to. Right now it's more of a meta id.

File: assets/js/translator.js [hunk 2]
```diff
@@ -868,9 +874,54 @@ function autoSelectDstLang() {
...
 
+function dictionaryLookup() {
+    if(modeEnabled('lookup')) {
+        if(currentLookupRequest) {
+            currentLookupRequest.abort();
+        }
+
+        $('#dictionaryLookupResult').empty();
+        var wordToLookup = $('#originalText').val().trim()
+            .split(' ');
```
Comment (sushain97) [inline assets/js/translator.js]: Suggest just putting this on one line and ignoring the rule with a comment.

File: assets/js/translator.js [hunk 3]
```diff
@@ -868,9 +874,54 @@ function autoSelectDstLang() {
...
+        if(wordToLookup.length <= 3 && wordToLookup[0] !== '') {
```
Comment (sushain97) [inline assets/js/translator.js]: Why 3?

File: assets/js/translator.js [hunk 4]
```diff
@@ -868,9 +874,54 @@ function autoSelectDstLang() {
...
+
+    function handleDictionaryLookupSuccessResponse(data) {
+        $('#dictionaryLookup').removeClass('hide');
+        $('#currentWordForLookup').html($('#originalText').val());
+        for(var analysisType in data) {
+            var values = '<ul id="dictionaryTranslationContents">';
+            for(var translationContent = 0; translationContent < data[analysisType].length; translationContent++) {
+                values += '<li>' + data[analysisType][translationContent].replace('#', '') + '</li>';
+            }
+            values += '</ul>';
```
Comment (sushain97) [inline assets/js/translator.js]: Could you make a `<ul>` first and then append list elements to it? String concatenation is not ideal.

File: assets/js/translator.js [hunk 5]
```diff
@@ -868,9 +874,54 @@ function autoSelectDstLang() {
...
+            $('#dictionaryLookupResult').html($('#dictionaryLookupResult').html() +
```
Comment (sushain97) [inline assets/js/translator.js]: Same concept here. The way you're constructing this stuff is a bit non-standard.

File: assets/strings/eng.json [hunk 6]
```diff
@@ -63,5 +63,27 @@
...
     "Install_Apertium": "Install Apertium",
-    "Install_Apertium_Para": "Are you experiencing slow responses? Our servers might be overloaded. Learn how to <a href='http://wiki.apertium.org/wiki/Installation' target='_blank'>install Apertium locally</a>."
+    "Install_Apertium_Para": "Are you experiencing slow responses? Our servers might be overloaded. Learn how to <a href='http://wiki.apertium.org/wiki/Installation' target='_blank'>install Apertium locally</a>.",
+    "Dictionary_translations": "Dictionary translations of ",
+    "n": "Noun",
```
Comment (sushain97) [inline assets/strings/eng.json]: I think these should be prefixed with something, e.g. `POS_n`.

File: assets/strings/eng.json [hunk 7]
```diff
@@ -63,5 +63,27 @@
...
     "Install_Apertium": "Install Apertium",
-    "Install_Apertium_Para": "Are you experiencing slow responses? Our servers might be overloaded. Learn how to <a href='http://wiki.apertium.org/wiki/Installation' target='_blank'>install Apertium locally</a>."
+    "Install_Apertium_Para": "Are you experiencing slow responses? Our servers might be overloaded. Learn how to <a href='http://wiki.apertium.org/wiki/Installation' target='_blank'>install Apertium locally</a>.",
+    "Dictionary_translations": "Dictionary translations of ",
```
Comment (sushain97) [inline assets/strings/eng.json]: This doesn't work for similar reasons to what I mentioned regarding displaying the APy endpoint.

Comment (share-with-me) [inline assets/strings/eng.json]: Yes. I shall change it now.

**Comment by sushain97** [comment]
- [ ] switch output to left side
- [ ] better detection of multiwords
- [ ] single request

Make issues for future:

- [ ] grammatical information (e.g. noun gender)
- [ ] translations back to English per sense(helps figure out which one is correct, highlighting)
- [ ] ranking of translations

**Comment by share-with-me** [comment]
>  switch output to left side

By left, do you mean on the right below `translatedText`? Currently, the output is on the left.

**Comment by sushain97** [comment]
Sorry, I meant switch to the right. My fault.

**Comment by Androbin** [comment]
I am looking into this as part of [this GCI Task](https://codein.withgoogle.com/dashboard/task-instances/4781648174383104/).
But for some reason `www.apertium.org/apy/dictionaryLookup` gives me a 404 response code no matter what.

**Comment by Stevenjin8** [comment]
Hey what needs to be worked on?

**Comment by Stevenjin8** [comment]
As in where are the conflicts?

**Comment by shardulc** [comment]
@Stevenjin8 If you checkout this branch locally and try to merge it into `master`, you will encounter conflicts because the code in `master` has changed.

**Comment by Stevenjin8** [comment]
@shardulc I can't find the url for this branch

**Comment by sushain97** [comment]
@Stevenjin8 this page is the branch?

**Comment by Stevenjin8** [comment]
Im a little confused as to what to do. I have cloned share-with-me/gsoc-mono-dictionary-lookup but whenever I do `git branch` I get `*master`. I'll go on the IRC later.

**Comment by sushain97** [comment]
@Stevenjin8 what about `git checkout gsoc-mono-dictionary-lookup`

**Comment by Stevenjin8** [comment]
ahh there we go, thanks

**Comment by JPJPJPOPOP** [comment]
@sushain97 
Could you explain more about what you mean about each of the following?

- better detection of multiwords (use analyze to get # of multiwords for inputs with < 10 spaces)
- single request
- localization of "Dictionary translations of light"

File: assets/js/translator.js (hunk 3 above)
Comment (JPJPJPOPOP) [reply assets/js/translator.js]: @sushain97 
I understand that this limits dictionary lookups for sentences with less than or equal to 3 words, but I don't know why we need this. Could I just remove it?

**Comment by sushain97** [comment]
> single request

There are currently multiple APy HTTP requests firing for one discrete dictionary step, if I remember correctly. That should be pared down to 1 somehow (might require changes to APy).

> localization of "Dictionary translations of light"

This sentence in the interface needs to be localized so that the "dictionary translations of" part is e.g. Spanish if the user sets the interface to Spanish. See the existing `dynamicLocalizations` code.

> better detection of multiwords (use analyze to get # of multiwords for inputs with < 10 spaces)

This limitation exists because it doesn't make sense to display dictionary translations when the user types in 50 words. See how Google translate does it for example. This particular request is a bit complex. I suggest reading about Apertium's analysis and multiwords. @jonorthwash might be able to explain it more succinctly than I.

Either way, I suggest picking a subset of the issues that I've brought up in comments and my review, fixing those and then asking if that seems to be enough for this task ;) There's a lot of stuff and I don't want you to be working on every single fix for days.

**Comment by JPJPJPOPOP** [comment]
@sushain97 

For the localization of "Dictionary translations of light", could I just translate "Dictionary translations of" into any one language? Also, I really don't know any 2nd language well enough, so can I use google translate?

File: assets/js/translator.js (hunk 3 above)
Comment (shardulc) [reply assets/js/translator.js]: @JPJPJPOPOP If you enter a sentence, you obviously want the entire sentence translation, not dictionary translations for each word. Where to draw the line is the question.

**Comment by shardulc** [comment]
@JPJPJPOPOP 

> For the localization of "Dictionary translations of light"

As discussed on IRC, "Dictionary translations of " + X makes sense in English but not all languages, so the phrase must be complete-ish in its meaning. "Dictionary translations for input:" is an option. If you don't know any other languages, don't translate it for now, it is preferable to have them translated by native/fluent speakers.

**Pull request closed without merging**